File: go/internal/cache/cache_test.go
package cache_test

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"
	"time"

	"codex-web-mcp/internal/cache"
)

func cacheFilePath(t *testing.T, dir, tool, payload string) string {
	t.Helper()
	return filepath.Join(dir, "cache", fmt.Sprintf("%s-%s.json", tool, cache.KeyHash(tool, payload)))
}

func TestKeyHash_Deterministic(t *testing.T) {
	a := cache.KeyHash("web_search", "hello world")
	b := cache.KeyHash("web_search", "hello world")
	if a != b {
		t.Fatalf("hash not deterministic: %s vs %s", a, b)
	}
	if len(a) != 32 {
		t.Fatalf("hash length should be 32, got %d", len(a))
	}
}

func TestKeyHash_DiffersByTool(t *testing.T) {
	a := cache.KeyHash("web_search", "x")
	b := cache.KeyHash("web_search_raw", "x")
	if a == b {
		t.Fatalf("hashes should differ across tools, got %s", a)
	}
}

func TestKeyHash_DiffersByPayload(t *testing.T) {
	a := cache.KeyHash("web_search", "x")
	b := cache.KeyHash("web_search", "y")
	if a == b {
		t.Fatalf("hashes should differ across payloads, got %s", a)
	}
}

func TestRoundTrip(t *testing.T) {
	dir := t.TempDir()
	t.Setenv("CODEX_WEB_CACHE", dir)

	cache.Put("web_search", "query1", "the answer")
	got, ok := cache.Get("web_search", "query1", 3600)
	if !ok {
		t.Fatalf("expected ok=true, got false")
	}
	if got != "the answer" {
		t.Fatalf("expected 'the answer', got %q", got)
	}
}

func TestGet_Miss(t *testing.T) {
	dir := t.TempDir()
	t.Setenv("CODEX_WEB_CACHE", dir)

	got, ok := cache.Get("web_search", "never-written", 3600)
	if ok {
		t.Fatalf("expected ok=false, got true with %q", got)
	}
	if got != "" {
		t.Fatalf("expected empty string, got %q", got)
	}
}

func TestGet_Expired(t *testing.T) {
	dir := t.TempDir()
	t.Setenv("CODEX_WEB_CACHE", dir)

	cache.Put("web_search", "stale", "old")
	p := cacheFilePath(t, dir, "web_search", "stale")
	past := time.Now().Add(-2 * time.Hour)
	if err := os.Chtimes(p, past, past); err != nil {
		t.Fatalf("chtimes: %v", err)
	}

	got, ok := cache.Get("web_search", "stale", 60)
	if ok {
		t.Fatalf("expected ok=false (expired), got true with %q", got)
	}
	if got != "" {
		t.Fatalf("expected empty string, got %q", got)
	}
}

func TestGet_WithinTTL(t *testing.T) {
	dir := t.TempDir()
	t.Setenv("CODEX_WEB_CACHE", dir)

	cache.Put("web_search", "fresh", "content")
	got, ok := cache.Get("web_search", "fresh", 3600)
	if !ok {
		t.Fatalf("expected ok=true within TTL, got false")
	}
	if got != "content" {
		t.Fatalf("expected 'content', got %q", got)
	}
}

File: node/tests/cache.test.ts
import { test } from "node:test";
import * as assert from "node:assert/strict";
import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import * as crypto from "node:crypto";

function withTempCache(fn: (dir: string) => void): void {
  const dir = path.join(os.tmpdir(), `codex-web-mcp-tests-${crypto.randomUUID().replace(/-/g, "")}`);
  fs.mkdirSync(dir, { recursive: true });
  const prev = process.env.CODEX_WEB_CACHE;
  process.env.CODEX_WEB_CACHE = dir;
  try {
    // Force re-import so the cache module reads the fresh env.
    fn(dir);
  } finally {
    if (prev === undefined) delete process.env.CODEX_WEB_CACHE;
    else process.env.CODEX_WEB_CACHE = prev;
    try {
      fs.rmSync(dir, { recursive: true, force: true });
    } catch {
      /* ignore */
    }
  }
}

test("keyHash is deterministic and 32 hex chars", async () => {
  const { keyHash } = await import("../src/cache.js");
  const a = keyHash("web_search", "hello world");
  const b = keyHash("web_search", "hello world");
  assert.equal(a, b);
  assert.equal(a.length, 32);
});

test("keyHash differs by tool", async () => {
  const { keyHash } = await import("../src/cache.js");
  assert.notEqual(keyHash("web_search", "x"), keyHash("web_search_raw", "x"));
});

test("keyHash differs by payload", async () => {
  const { keyHash } = await import("../src/cache.js");
  assert.notEqual(keyHash("web_search", "x"), keyHash("web_search", "y"));
});

test("round trip", async () => {
  const cache = await import("../src/cache.js");
  withTempCache(() => {
    cache.put("web_search", "query1", "the answer");
    assert.equal(cache.get("web_search", "query1", 3600), "the answer");
  });
});

test("miss returns null", async () => {
  const cache = await import("../src/cache.js");
  withTempCache(() => {
    assert.equal(cache.get("web_search", "never-written", 3600), null);
  });
});

test("expired returns null", async () => {
  const cache = await import("../src/cache.js");
  withTempCache((dir) => {
    cache.put("web_search", "stale", "old");
    const file = path.join(dir, "cache", `web_search-${cache.keyHash("web_search", "stale")}.json`);
    const past = (Date.now() - 2 * 3600 * 1000) / 1000;
    fs.utimesSync(file, past, past);
    assert.equal(cache.get("web_search", "stale", 60), null);
  });
});

test("within TTL returns content", async () => {
  const cache = await import("../src/cache.js");
  withTempCache(() => {
    cache.put("web_search", "fresh", "content");
    assert.equal(cache.get("web_search", "fresh", 3600), "content");
  });
});
